modules/mongodb/templates/mongodb-backup-s3.erb

```diff
@@ -17,65 +17,100 @@ date_format = '%Y-%m-%d:%H:%M:%S'
-gpg.communicate()
-
-logger('End compress and encrypt backups')
-
-s3 = boto3.resource('s3')
-
-logger('Start upload file to S3')
-
-s3_file = backup_dir + '/' + file
-s3.Object(s3_bucket, s3_file).put(Body=open(s3_file, 'rb'))
-
-logger('End upload file to S3')
-
-logger('Start cleanup filesystem')
-
-delete_tree = backup_dir + '/*'
-[os.remove(os.path.join(backup_dir,f)) for f in os.listdir(backup_dir) if f.endswith(".gpg")]
-shutil.rmtree(delete_tree)
-
-logger('End cleanup filesystem')
-
-end_time = time.time()
-delta = end_time - start_time
-
-print 'Script completed in     : %.1f seconds' % delta
+def get_mongo_nodes():
+    logger('Find Secondary Mongo members')
+    try:
+        mongo_nodes = subprocess.check_output(["mongo",  mongo_server, "--quiet", "--eval", "printjson(rs.status().members.map(function(m) { return {'name':m.name, 'stateStr':m.stateStr} }))"])
+    except OSError:
+        sys.exit('Trouble querying mongo server: %s' % mongo_server)
+
+    nodes = json.loads(mongo_nodes)
+    secondary_nodes = { x["name"] for x in nodes if x["stateStr"] == "SECONDARY" }
+
+    if secondary_nodes:
+      backup_nodes = random.sample(secondary_nodes,1)
+    else:
+      backup_nodes = ['localhost']
+
+    backup_node = backup_nodes[0]
+    return backup_node
+
+def backup_mongo(node):
+    logger('Start mongo backup')
+    try:
+        subprocess.call(['mongodump', '-h', node, '-o', backup_dir])
+    except OSError:
+        sys.exit('There was a problem backing up the mongo database')
+    logger('End mongo backup')
+
+def get_subdirs(parent):
+    subdirs = os.listdir(parent)
+    return subdirs
+
+def compress_and_encrypt(backup_dir, dir):
+    logger('Start compress and encrypt backups for %s' % dir)
+    subdir = backup_dir + '/' + str(dir)
+    enc_file = backup_dir + '/' + str(dir) + '.tar.gz.gpg'
+    try:
+        tar = subprocess.Popen(['tar', '-cvz', subdir], stdout=subprocess.PIPE)
+        gpg = subprocess.Popen(['gpg', '-e', '-o', str(enc_file) , '-r', key_fingerprint, '--cipher-algo', 'AES256', '--force-mdc'], stdin=tar.stdout, stdout=subprocess.PIPE)
+        tar.stdout.close()
+        gpg.communicate()
+    except OSError as e:
+        sys.exit('There was a problem compressing and encrypting the database')
+    logger('End compress and encrypt backups for %s' % dir)
+
+def push_to_s3(backup_dir, dir):
+    try:
+        enc_file = str(dir) + '.tar.gz.gpg'
+        logger('Start upload file %s to S3' % enc_file)
+        s3 = boto3.resource('s3')
+        s3_key = str(today) + '/' + str(enc_file)
+        source = backup_dir + '/' + str(enc_file)
+        s3.Object(s3_bucket, s3_key).put(Body=open(source, 'rb'))
+        logger('End upload file %s to S3' % enc_file)
+    except Exception as e:
+        sys.exit(e)
+
+def cleanup_local():
+    logger('Start cleanup filesystem')
+    if backup_dir not in ['','/']:
+        rm_tree = backup_dir + '/'
+        try:
+            subprocess.call(['rm', '-rf', rm_tree])
```

Comment:
This is pretty scary - it feels like there's quite a bit of potential for this to be misconfigured and remove something we care about.

Could we do something else instead? Given that this script is the same script that pushes the file to S3 would it be possible to remove the exact file we've pushed immediately after we push it?